File: python/src/hearth/bip39.py
"""BIP-39 mnemonic validation and seed derivation."""

from __future__ import annotations

import hashlib
import unicodedata
from functools import cache
from importlib.resources import files

_ITERATIONS = 2048
_SEED_LEN = 64
_VALID_WORD_COUNTS = {12, 15, 18, 21, 24}


@cache
def wordlist() -> tuple[str, ...]:
    text = files("hearth.data").joinpath("english.txt").read_text(encoding="utf-8")
    return tuple(line.strip() for line in text.splitlines() if line.strip())


@cache
def _word_index() -> dict[str, int]:
    return {word: i for i, word in enumerate(wordlist())}


def _normalize(text: str) -> str:
    return unicodedata.normalize("NFKD", text)


def validate(mnemonic: str) -> str | None:
    """Return an error message if invalid, else ``None`` (valid checksum)."""
    words = [w for w in _normalize(mnemonic).split(" ") if w]
    if len(words) not in _VALID_WORD_COUNTS:
        return f"word count must be 12/15/18/21/24, got {len(words)}"
    index = _word_index()
    indices: list[int] = []
    for word in words:
        if word not in index:
            return f"unknown word: '{word}'"
        indices.append(index[word])

    bits = [(i >> b) & 1 for i in indices for b in range(10, -1, -1)]
    checksum_bits = len(bits) // 33
    entropy_bits = len(bits) - checksum_bits
    entropy = _bits_to_bytes(bits[:entropy_bits])
    digest = hashlib.sha256(entropy).digest()
    expected = [(digest[i // 8] >> (7 - (i % 8))) & 1 for i in range(checksum_bits)]
    if expected != bits[entropy_bits:]:
        return "checksum mismatch"
    return None


def to_seed(mnemonic: str, passphrase: str = "") -> bytes:
    """Derive the 64-byte BIP-39 seed (PBKDF2-HMAC-SHA512, 2048 iterations)."""
    password = _normalize(mnemonic).encode("utf-8")
    salt = _normalize("mnemonic" + passphrase).encode("utf-8")
    return hashlib.pbkdf2_hmac("sha512", password, salt, _ITERATIONS, _SEED_LEN)


def _bits_to_bytes(bits: list[int]) -> bytes:
    assert len(bits) % 8 == 0
    out = bytearray(len(bits) // 8)
    for i in range(0, len(bits), 8):
        byte = 0
        for b in bits[i : i + 8]:
            byte = (byte << 1) | b
        out[i // 8] = byte
    return bytes(out)

File: go/hearth/bip39.go
package hearth

import (
	"crypto/pbkdf2"
	"crypto/sha256"
	"crypto/sha512"
	_ "embed"
	"errors"
	"fmt"
	"strings"

	"golang.org/x/text/unicode/norm"
)

//go:embed english.txt
var englishWordlist string

var (
	wordlist  []string
	wordIndex map[string]int
)

func init() {
	wordlist = strings.Fields(englishWordlist)
	wordIndex = make(map[string]int, len(wordlist))
	for i, w := range wordlist {
		wordIndex[w] = i
	}
}

func normalizeNFKD(s string) string { return norm.NFKD.String(s) }

// ValidateMnemonic returns nil for a valid BIP-39 mnemonic, else an error.
func ValidateMnemonic(mnemonic string) error {
	words := strings.Fields(normalizeNFKD(mnemonic))
	n := len(words)
	if n != 12 && n != 15 && n != 18 && n != 21 && n != 24 {
		return fmt.Errorf("word count must be 12/15/18/21/24, got %d", n)
	}
	indices := make([]int, n)
	for i, w := range words {
		idx, ok := wordIndex[w]
		if !ok {
			return fmt.Errorf("unknown word: %q", w)
		}
		indices[i] = idx
	}

	totalBits := n * 11
	checksumBits := totalBits / 33
	entropyBits := totalBits - checksumBits
	bits := make([]int, 0, totalBits)
	for _, idx := range indices {
		for b := 10; b >= 0; b-- {
			bits = append(bits, (idx>>b)&1)
		}
	}
	digest := sha256.Sum256(bitsToBytes(bits[:entropyBits]))
	for i := 0; i < checksumBits; i++ {
		expected := int((digest[i/8] >> (7 - (i % 8))) & 1)
		if bits[entropyBits+i] != expected {
			return errors.New("checksum mismatch")
		}
	}
	return nil
}

// MnemonicToSeed derives the 64-byte BIP-39 seed (PBKDF2-HMAC-SHA512, 2048 iters).
func MnemonicToSeed(mnemonic, passphrase string) []byte {
	password := normalizeNFKD(mnemonic)
	salt := normalizeNFKD("mnemonic" + passphrase)
	seed, err := pbkdf2.Key(sha512.New, password, []byte(salt), 2048, 64)
	if err != nil {
		panic(err)
	}
	return seed
}

func bitsToBytes(bits []int) []byte {
	out := make([]byte, len(bits)/8)
	for i := 0; i < len(bits); i += 8 {
		var b int
		for _, bit := range bits[i : i+8] {
			b = (b << 1) | bit
		}
		out[i/8] = byte(b)
	}
	return out
}
